design/hostclaim-multitenancy-and-hybrid-clusters.md

```diff
@@ -0,0 +1,678 @@
+for a recyclable resource (such as a bare-metal server) or as characteristics
+for a disposable resource (such as a virtual machine created for the workload).
+The status and meta-data of the HostClaim provide the necessary information
+for the end user to define and manage his workload on the compute resource,
+but they do not grant full control over the resource (typically, BMC
+credentials of servers are not exposed to the tenant).
+
+## Motivation
+
+The standard approach to implementing multi-tenancy in cluster-api is to follow the
+[multi tenancy contract](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/multi-tenancy#contract).
+
+To adhere to this contract with cluster-api-provider-metal3, the clusters must
+be put in different namespaces and the BareMetalHost objects must be defined
+in those namespaces. In this setup, the tenants are the owners of the servers
+and it becomes difficult to share the same server between different clusters if
+they belong to different tenants.
+
+In order to improve server usage, we would like to have a pool of servers that
+clusters can lease depending on their workload. If we maintain the Metal3
+constraints, all clusters must be defined in the same namespace. Unless very
+complex access control rules are defined, cluster administrators have
+visibility and probably control over all clusters and servers as the server
+credentials are stored with the BareMetalHost resource.
+
+We need to relax the constraint that the cluster and the BareMetalHosts are
+in the same namespace but we also need a solution that give sufficient control
+and visibility over the workload deployed on those servers so that tenants
+can maintain the level of information they have had so far.
+
+Another unrelated problem is that Cluster-API has been designed
+to define clusters using homogeneous compute resources: it is challenging to
+define a cluster with both bare-metal servers and virtual machines in a private
+or public cloud.
+This [blog post](https://metal3.io/blog/2022/07/08/One_cluster_multiple_providers.html)
+proposes several approaches but none is entirely satisfactory.
+
+On the other hand, workloads can be easily defined in terms of OS images and
+initial configurations and standards such as qcow or cloud-init have emerged
+and are used by various infrastructure technologies.
+Due to the unique challenges of managing bare-metal, the Metal3 project has
+developed a set of abstractions and tools that could be used in different
+settings. The main mechanism is the selection process that occurs between
+the Metal3Machine and the BareMetalHost which assigns a disposable workload
+(being a Kubernetes node) to a recyclable compute resource (a
+server).
+
+This proposal introduces a new resource called HostClaim that solves
+both problems by decoupling the definition of the workload performed by
+the Metal3 machine controller from the actual compute resource.
```

Comment:
Do you mean access to the BMC ? It must not be given to tenants. In the same way that they must not see the BMC credentials. An important note is that if you pivot a cluster with hostclaim, you are not pivoting ironic (unless everybody is in the same namespace and this is essentially for bootstrap). Because bootstrap is a special case, there is a dedicated scenario that tells essentially that for bootstrap, the old behaviour is kept because everything is moved.

Do you mean network isolation between servers: maintaining it is outside the scope of this proposal. I think DT has a proposal where each cluster ensure its own isolation through "signed BGP" advertisment. Somehow the newly booted server comes with some credentials that make that it will have access to/publish some routes. In our integration project, we will probably try to make an evolution of the networking solution we had with Kanod baremetalpools: https://orange-opensource.gitlab.io/kanod/reference/kiab/network.html#kiab-network. But I think it is safe to say that it is outside the scope of Kanod.